// components/Messages/ChatList.jsx
import React, {Component} from 'react';
import { ChatList } from 'react-chat-elements';

export default function List() {
    return (
        <ChatList
            className='chat-list'
            dataSource={[
                {
                    avatar: 'https://image.flaticon.com/icons/svg/701/701993.svg',
                    alt: 'Reactjs',
                    title: 'Facebook',
                    subtitle: 'What are you doing?',
                    date: new Date(),
                    unread: 0,
                },
                {
                    avatar: 'https://image.flaticon.com/icons/svg/701/701993.svg',
                    alt: 'Reactjs',
                    title: 'Facebook',
                    subtitle: 'What are you doing?',
                    date: new Date(),
                    unread: 0,
                },
                {
                    avatar: 'https://image.flaticon.com/icons/svg/701/701993.svg',
                    alt: 'Reactjs',
                    title: 'Facebook',
                    subtitle: 'What are you doing?',
                    date: new Date(),
                    unread: 0,
                }
            ]}
        />
    )
}
