// components/MessageList/index.jsx
import React, { Fragment } from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';

//Components
import MessageList from './MessageList';

//HOCS
import loadMessages from './loadMessages'
import {loading} from 'components/HOC/loading';

import { GetMessageList } from 'actions';

class MessagesPage extends React.Component {

    renderMessageList = (list) => {
        return list.map((item, index) => {
            let checkLastMessage = item.last_message;
            return {
                id: item.id,
                avatar: 'https://image.flaticon.com/icons/svg/701/701993.svg',
                title: item.first_user.username,
                subtitle: checkLastMessage ? item.last_message.text : 'Empty chat',
                dateString: checkLastMessage ? item.last_message.created_at : ''
            }
        })
    }

    render() {
        const { messageList } = this.props;
        return (
            <Fragment>
                <MessageList messageList={messageList}/>
            </Fragment>
        );
    }
}

const MapStateToProps = (state) => ({
    messageList: state.messageList
})

export default compose(
    connect( MapStateToProps, { GetMessageList }),
    loadMessages,
    loading
)(MessagesPage)
