// actions/chat.js
import {
    CHANGE_CHAT_LAST_MESSAGE, CHANGE_CURRENT_CHAT, FETCH_CHATS, FETCH_CHATS_FAILURE,
    FETCH_CHATS_SUCCESS
} from "constants/action-types";
import {API_URL} from "constants/server";
import axios from 'axios'

const changeCurrentChat = chat_id => ({type: CHANGE_CURRENT_CHAT, payload: chat_id});
const changeChatLastMessage = (last_message) => ({
    type: CHANGE_CHAT_LAST_MESSAGE,
    payload: last_message
});

function fetchChats() {
    const token = localStorage.getItem('token');
    const config = {
        headers: {'Authorization': 'JWT ' + token}
    };
    const request = axios.get(`${API_URL}/private-chats/`, config);
    return {
        type: FETCH_CHATS,
        payload: request
    }

}

function fetchChatsSuccess(chats) {
    return {
        type: FETCH_CHATS_SUCCESS,
        payload: chats
    }
}

function fetchChatsFailure(error) {
    return {
        type: FETCH_CHATS_FAILURE,
        payload: error
    }
}

export {changeCurrentChat, changeChatLastMessage, fetchChats, fetchChatsSuccess, fetchChatsFailure}
