// components/ChatList/index.jsx
import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

//Components
import ChatList from './ChatList';

//HOCS
import { loadChat } from 'components/HOCs';

import { GetChatList } from 'actions';

class ChatsPage extends React.Component {
    constructor() {
        super();
        this.moveToDirectChat = this.moveToDirectChat.bind(this);
    }

    renderChatList = (list) => {
        return list.map((item, index) => {
            let checkLastMessage = item.last_message;
            return {
                id: item.id,
                avatar: 'https://image.flaticon.com/icons/svg/701/701993.svg',
                title: item.first_user.username,
                subtitle: checkLastMessage ? item.last_message.text : 'Empty chat',
                dateString: checkLastMessage ? item.last_message.created_at : ''
            }
        })
    }

    moveToDirectChat(item){
        this.props.history.push(`/messages/${item.id}`)
    }

    render() {
        const { chatList } = this.props;
        return (
            <ChatList 
                chatList={chatList}
                renderChatList={this.renderChatList}
                moveToDirectChat={this.moveToDirectChat}
            />
        );
    }
}

const MapStateToProps = (state) => ({
    chatList: state.chatList
})

export default compose(
    connect( MapStateToProps, { GetChatList }),
    loadChat
)(withRouter(ChatsPage))
