// constants/index.js
export const GET_USER_INFO = 'GET_USER_INFO';

export const REGISTER = 'REGISTER';
export const LOGIN = 'LOGIN';

export const user = {
    REGISTER_FAILURE: 'REGISTER_FAILURE',
    REGISTER_SUCCESS: 'REGISTER_SUCCESS',
    REGISTER_LOAD: 'REGISTER_LOAD',

    LOGIN_FAILURE: 'LOGIN_FAILURE',
    LOGIN_SUCCESS: 'LOGIN_SUCCESS',
    LOGIN_LOAD: 'LOGIN_LOAD',
    LOGOUT: 'LOGOUT',
    LOGIN_USER_ID: 'LOGIN_USER_ID'
};

export const chat = {
    GET_CHAT_LIST: 'GET_CHAT_LIST',
    CLEAR_CHAT_LIST: 'CLEAR_MESSAGE_LIST'
}

export const message = {
    GET_MESSAGE_LIST: 'GET_MESSAGE_LIST',
    RECEIVE_MESSAGE: 'RECEIVE_MESSAGE',
    CLEAR_MESSAGE_LIST: 'CLEAR_MESSAGE_LIST'
}

export const messageSocket = {
    TEXT_MESSAGE: 'chat-text-message',
}

export const CHAT_TEXT_MESSAGE = 'chat-text-message';
export const CHAT_FILE_MESSAGE = 'chat-file-message';
export const CHAT_IMAGE_MESSAGE = 'chat-image-message';
export const CHAT_AUDIO_MESSAGE = 'chat-audio-message';
export const CHAT_VIDEO_MESSAGE = 'chat-video-message';

export const CHAT_UPDATE_TEXT_MESSAGE = 'chat-update-text-message';


export const CHAT_DELETE_TEXT_MESSAGE = 'chat-delete-text-message';
export const CHAT_DELETE_FILE_MESSAGE = 'chat-delete-file-message;';
export const CHAT_DELETE_IMAGE_MESSAGE = 'chat-delete-image-message';
export const CHAT_DELETE_AUDIO_MESSAGE = 'chat-delete-audio-message';
export const CHAT_DELETE_VIDEO_MESSAGE = 'chat-delete-video-message';
