// routers/index.jsx
import React from 'react';
import { BrowserRouter as Router, Route, Switch, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { compose } from 'redux';

//Routes
import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';

//HOCs
import { subscribeNotifies, subscribeMessages, loadUserInfo } from 'components/HOCs';

//Components
import {
    MainPage,
    Login,
    Register,
    NoMatch,
    MessageList,
    ChatList
} from 'components';

import localStorageApi from 'services/localStorageApi';
import { isSuccess, isFailure, GetUserID } from 'actions';
import { LOGIN } from '../constants';

class Routes extends React.Component {

    componentDidMount(token = localStorageApi.getItem()) {
        this.props.setLoggedIn({token});
        this.props.subscribeNotifies(token);
        this.props.loadUserInfo(token);
    }

    render() {
        return(
            <Router>
                <Switch>
                    {/* LoggedIn Routes */}
                    <PrivateRoute exact path="/" component={MainPage}/>
                    <PrivateRoute 
                        path='/messages/:id' 
                        routeProps={
                            {
                                subscribeMessages: this.props.subscribeMessages,
                                sendMessage: this.props.sendMessage
                            }
                        }
                        component={MessageList}
                    />
                    <PrivateRoute path='/messages' component={ChatList}/>
                    
                    {/* LoggedOut Routes */}
                    <PublicRoute 
                        exact path="/login"
                        routeProps={
                            {
                                subscribeNotifies: this.props.subscribeNotifies,
                                loadUserInfo: this.props.loadUserInfo
                            }
                        } 
                        component={Login}
                    />
                    <PublicRoute path="/register" component={Register} />

                    {/* Other Routes */}
                    <Route
                        path="/logout"
                        render={() => {
                            localStorageApi.clear();
                            return <Redirect to="login"/>;
                        }}
                    />
                    <Route exact path="*" render={() => <NoMatch isLoggedIn={this.props.isLoggedIn}/>} />
                </Switch>
            </Router>
       )
    }
}

const mapStateToProps = state => (
    {
        isLoggedIn: state.login.isLoggedIn
    }
);

const mapDispatchToProps = dispatch => (
    {
        setLoggedIn: payload => payload ? dispatch(isSuccess(LOGIN, payload)) : dispatch(isFailure(LOGIN)),
        GetUserID: payload => dispatch(GetUserID(payload))
    }
);

export default compose(
    connect(mapStateToProps, mapDispatchToProps),
    loadUserInfo,
    subscribeMessages,
    subscribeNotifies
)(Routes);
