// components/HOCs/subscribeMessages.jsx
import React from 'react';

import { messageSocket } from "constants/index";

import { SocketApi } from 'services';

const socketSubscribe = (Component) => {
    class AsyncComponent extends React.Component {
        constructor() {
            super();
            this.subscribeMessages = this.subscribeMessages.bind(this);
            this.sendMessage = this.sendMessage.bind(this);
        }

        subscribeMessages(token, chatId) {
            if (!token && !chatId) console.log('Something wasn\'t match');
            this.socket = new SocketApi(
                `ws://0.0.0.0:8000/private_chat/?token=${token}&private_chat_id=${chatId}`
            )
            this.receiveMessage();

            //Delete in PROD!
            this.socket.onopen = () => {
                console.log("User message_socket open");
            };
            this.socket.onclose = () => {
                console.log('User message_socket disconnected')
            };
            this.socket.onerror = (e) => {
                console.log(e)
            };
            //-------------------
        }

        receiveMessage() {
            this.socket.onmessage = (event) => {
                console.log(event);
                const data = JSON.parse(event.data);
                //chat-text-message
                if (data.action_type === messageSocket.TEXT_MESSAGE) {
                    this.props.ReceiveMessage(data);
                }
            };
        }

        sendMessage(message) {
            this.socket.send(JSON.stringify(message));
        }
    
        render() {
            return (
                <Component
                    subscribeMessages={this.subscribeMessages}
                    sendMessage={this.sendMessage}
                    {...this.props} 
                />
            )
        }
    }

    return AsyncComponent;
};


export default socketSubscribe;
